force-app/main/default/classes/DMLOperations.cls

```diff
@@ -180,15 +205,13 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
+		Account contactAccount; // Create null instance of Account
+        for( Contact c : contactsToUpsert) { // Loop through each Contact
+            contactAccount = upsertAccount(c.LastName); // Extract the Account Name from Contact's LastName and call previous question to upsert Account
```

Comment:
Your upsertAccount method does SOQL and DML, so calling that from a loop is still DML in a loop, even though the PMD checks dont catch it. It was a good thought to call another method that did this, but that method cant be called from inside a loop.